### DISCUSSION TITLE
FromMemory / FromStream accross the board

### DISCUSSION THREAD

**Discussion by llfab** [StereoKit/StereoKit#300] [Ideas]
Luckily, many scene element provide a .FromFile AND a .FromMemory method. However, this seems not there accross the board. E.g., `Sprite` is missing such a function and can only be retrieved from memory through `Tex`. 

We could add .FromMemory accross all elements that have a .FromFile method. However, another pattern could be worthwhile:
`Type.FromMemory(Stream stream, TypeFormat format, <additional params if needed>)`. This can very nicely be used by the Type.FromFile methods whereas .FromFile just retrieves the TypeFormat from the file extension and then calls the stream based function.

**Comment by maluoi** [comment]
Sprite is a little different, since it works with a Tex underneath one way or another. FromMemory bypasses "filename already loaded" checks, since there's no filename to look at, so asking the user to manage the Tex more directly can help prevent duplicate GPU resources. I'll think about it a bit though, it has always felt like a bit of a weird detail.

I've had the Streams thing on my list for a while! (#69) I would likely implement this as just something that converts to byte[] and call FromMemory with it. I just need to go do it :) That'd also bypass filename checks though, so there's still some wrinkles I need to dance around there.

Having a generic asset loading function is perhaps the wrong direction for StereoKit. Engines like Unity have .meta files that are associated with asset files and describe information and settings about how to load and parse a particular asset. Since SK doesn't have this, there's somewhat of a need to provide some of those extra details when loading! So there are plans for having something like a, TexImportSettings argument you would provide when loading an image, or a ModelImportSettings for Models. This would make the signatures less uniform than they are currently and prevent a generic asset loading function from being sufficiently robust,

**Reply by llfab** [reply]
Sounds good to me. Also: I of course did not mean a generic function. `Type.FromMemory` was rather meant as an example whereas Type is Model, Tex or something else. Naturally, it would only make sense to have specific functions for specific types. Sorry, I indeed did not describe that right. 
`Tex.FromMemory(Stream stream, ...)` and `Model.FromMemory()` and `Sprite.FromMemory()` (if the latter even makes sense) was really what I wanted to point out.

**Reply by maluoi** [reply]
Ah 😅 I get it, yeah! Definitely +1 for #69 then!
